README.rst
argparse: Command line optional and positional argument parser.
===============================================================

.. image:: https://www.r-pkg.org/badges/version/argparse
    :target: https://cran.r-project.org/package=argparse
    :alt: CRAN Status Badge

.. image:: https://travis-ci.org/trevorld/argparse.png?branch=master
    :target: https://travis-ci.org/trevorld/argparse
    :alt: Build Status

.. image:: https://img.shields.io/codecov/c/github/trevorld/argparse.svg
    :target: https://codecov.io/github/trevorld/argparse?branch=master
    :alt: Coverage Status

.. image:: https://cranlogs.r-pkg.org/badges/argparse
    :target: https://cran.r-project.org/package=argparse
    :alt: RStudio CRAN mirror downloads

A command line parser to
be used with Rscript to write "#!" shebang scripts that gracefully
accept positional and optional arguments and automatically generate usage.

To install the development version use the following command::

    devtools::install_github("argparse", "trevorld")

dependencies
============

The package has a Python dependency.  
It is easily satisfied if you have Python (version 3.2 or higher) on your PATH.
Read the INSTALL file for more information if this doesn't describe you.

Additionally this package depends on the R packages ``proto``, ``findpython``,
``getopt``, and ``jsonlite``.

To run the unit tests you will need the suggested R package ``testthat`` and in
order to build the vignette you will need the suggested R package ``knitr`` 
which in turn probably requires the system tool ``pandoc``::

    sudo apt-get install pandoc

example
=======

::

  > parser <- ArgumentParser(description='Process some integers')
  > parser$add_argument('integers', metavar='N', type="integer", nargs='+',
  +                    help='an integer for the accumulator')
  > parser$add_argument('--sum', dest='accumulate', action='store_const',
  +                    const='sum', default='max',
  +                    help='sum the integers (default: find the max)')
  > parser$print_help()
  usage: PROGRAM [-h] [--sum] N [N ...]
  
  Process some integers
  
  positional arguments:
    N           an integer for the accumulator
  
  optional arguments:
    -h, --help  show this help message and exit
    --sum       sum the integers (default: find the max)

Default args for ``ArgumentParser()$parse_args`` are ``commandArgs(TRUE)``
which is what you'd want for an Rscript but not for interactive use::

  > args <- parser$parse_args(c("--sum", "1", "2", "3")) 
  > accumulate_fn <- get(args$accumulate)
  > print(accumulate_fn(args$integers))
  [1] 6


NEWS
argparse 1.1.1
==============
* Better support for machines only running Python 2.7.  
* DESCRIPTION makes explicit that it requires version 1.0.0 of ``proto`` dependency.

argparse 1.1.0
==============
* Better support for Unicode arguments/options.
  Thanks Erick Rocha Fonseca for bug report.
* Fix bug when setting ``type='integer'`` in ``add_argument``.
  Thanks Dominik Mueller for bug report and suggested solution.
* The package ``proto`` has been moved from ``Depends`` to ``Imports`` in ``DESCRIPTION``.
  We now use ``jsonlite`` package instead of ``rjson`` to parse ``JSON``.
* Better error message when accidentally trying to define a positional argument in ``ArgumentParser()``
  Suggestion of Alex Reinhart.
* Other error message output when an parsing error is detected have been tweaked slightly.

argparse 1.0.7
==============
* Now supports showing program's version with ``action='version'``.
  Suggestion of Dario Beraldi.

argparse 1.0.6
==============
* Fix bug when setting ``required=FALSE`` in ``add_argument``.
  Thanks Claire D. McWhite for bug report.

argparse 1.0.4
==============
* You can now pass in a character vector to the ``metavar`` argument of ``add_argument``.
  Thanks Paul Newell for bug report and a patch.

argparse 1.0.3
==============
* `add_argument` now throws a warning recommending using action "store_true" or "store_false" 
  if type set to "logical" and action set to "store".  Suggestion of Martí Duran Ferrer
* You can now explicitly set a `NULL` default in `add_argument`.  
  Previously one could only implicitly set a `NULL` default by not setting any default at all.
  Suggestion of Hyunsoo Kim.

argparse 1.0.2
==============
* Fixes parsing bug when using a very large argument list.  Thanks Taylor Pospisil for bug report.
* Parse error usage message now prints to standard error.  User requested help message quit status is now zero.  
  Thanks to PlasmaBinturong for report/request.
* If Python script fails with error pass on error message to user.
  Thanks to Martí Duran Ferrer for report/request.

argparse 1.0.1
==============
* Fix minor bug when running package tests on machines with both old and new versions of Python.  Thanks Prof. Brian Ripley for bug report.

argparse 1.0.0
==============
* Some support for ``choices`` argument in ``parser$add_argument``.  Thanks Christopher Small for report/request.

* Some support for ``formatter_class`` argument in ``ArgumentParser``.  Thanks Costas Bouyioukos for report/request.

* Package now looks for Python binaries with ``findpython::find_python_cmd``
  which is a little bit more robust than the previous ``argparse::find_python_cmd``.

* If ``interactive() == TRUE`` then ``parse_args`` will no longer ``quit(status=1)`` after printing a help message
  but will instead throw an error.  ``argparse`` will continue to ``quit(status=1)`` after printing a help message
  for non-interactive Rscripts.


argparse 0.5.3
==============

* Fix bug in parser$add_argument for case when we specify a default vector
  instead of scalar.  Thanks Martin Diehl for bug report.

* ``find_python_cmd`` will now also look for a ``python2.7`` binary.

* Website moved to https://github.com/trevorld/argparse


tests/run-all.R
library("testthat")

test_check("argparse")


tests/test_help.R
library("argparse")
p <- ArgumentParser()
p$parse_args("--help")


tests/test_version.R
library("argparse")
p <- ArgumentParser()
p$add_argument("-v", "--version", action="version", version='1.0.1')
p$parse_args("--version")


exec/example.R
#!/usr/bin/env Rscript
# Copyright 2012-2013 Trevor L Davis <trevor.l.davis@stanford.edu>
# Copyright 2008 Allen Day
#  
#  This file is free software: you may copy, redistribute and/or modify it  
#  under the terms of the GNU General Public License as published by the  
#  Free Software Foundation, either version 2 of the License, or (at your  
#  option) any later version.  
#  
#  This file is distributed in the hope that it will be useful, but  
#  WITHOUT ANY WARRANTY; without even the implied warranty of  
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU  
#  General Public License for more details.  
#  
#  You should have received a copy of the GNU General Public License  
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.  
suppressPackageStartupMessages(library("argparse"))

# create parser object
parser <- ArgumentParser()

# specify our desired options 
# by default ArgumentParser will add an help option 
parser$add_argument("-v", "--verbose", action="store_true", default=TRUE,
    help="Print extra output [default]")
parser$add_argument("-q", "--quietly", action="store_false", 
    dest="verbose", help="Print little output")
parser$add_argument("-c", "--count", type="integer", default=5, 
    help="Number of random normals to generate [default %(default)s]",
    metavar="number")
parser$add_argument("--generator", default="rnorm", 
    help = "Function to generate random deviates [default \"%(default)s\"]")
parser$add_argument("--mean", default=0, type="double",
    help="Mean if generator == \"rnorm\" [default %(default)s]")
parser$add_argument("--sd", default=1, type="double",
        metavar="standard deviation",
    help="Standard deviation if generator == \"rnorm\" [default %(default)s]")
                                        
# get command line options, if help option encountered print help and exit,
# otherwise if options not found on command line then set defaults, 
args <- parser$parse_args()

# print some progress messages to stderr if "quietly" wasn't requested
if ( args$verbose ) { 
    write("writing some verbose output to standard error...\n", stderr()) 
}

# do some operations based on user input
if( args$generator == "rnorm") {
    cat(paste(rnorm(args$count, mean=args$mean, sd=args$sd), collapse="\n"))
} else {
    cat(paste(do.call(args$generator, list(args$count)), collapse="\n"))
}
cat("\n")


exec/display_file.R
#!/usr/bin/env Rscript
# Copyright 2012-2013 Trevor L Davis <trevor.l.davis@stanford.edu>
#  
#  This file is free software: you may copy, redistribute and/or modify it  
#  under the terms of the GNU General Public License as published by the  
#  Free Software Foundation, either version 2 of the License, or (at your  
#  option) any later version.  
#  
#  This file is distributed in the hope that it will be useful, but  
#  WITHOUT ANY WARRANTY; without even the implied warranty of  
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU  
#  General Public License for more details.  
#  
#  You should have received a copy of the GNU General Public License  
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.  
suppressPackageStartupMessages(library("argparse"))

parser <- ArgumentParser()
parser$add_argument("-n", "--add_numbers", action="store_true", default=FALSE,
    help="Print line number at the beginning of each line [default]")
parser$add_argument("file", nargs=1, help="File to be displayed")

args <- parser$parse_args()

file <- args$file

if( file.access(file) == -1) {
    stop(sprintf("Specified file ( %s ) does not exist", file))
} else {
    file_text <- readLines(file)
}

if(args$add_numbers) {
    cat(paste(1:length(file_text), file_text), sep = "\n")
} else {
    cat(file_text, sep = "\n")
}


R/argparse.R
# Copyright (c) 2012-2017 Trevor L. Davis <trevor.l.davis@stanford.edu>  
#  
#  This file is free software: you may copy, redistribute and/or modify it  
#  under the terms of the GNU General Public License as published by the  
#  Free Software Foundation, either version 2 of the License, or (at your  
#  option) any later version.  
#  
#  This file is distributed in the hope that it will be useful, but  
#  WITHOUT ANY WARRANTY; without even the implied warranty of  
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU  
#  General Public License for more details.  
#  
#  You should have received a copy of the GNU General Public License  
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.  
#  
# This file incorporates work from the argparse module in Python 2.7.3.
#  
#     Copyright (c) 2001-2012 Python Software Foundation; All Rights Reserved
#
# See (inst/)COPYRIGHTS or http://docs.python.org/2/license.html for the full
# Python (GPL-compatible) license stack.

#' Create a command line parser 
#'
#' \code{ArgumentParser} creates a parser object that acts as 
#' a wrapper to Python's argparse module
#' 
#' @param ... Arguments cleaned and passed to Pythons argparse.ArgumentParser()
#' @param python_cmd Python executable for \code{argparse} to use.
#'      Must have argparse and json modules (automatically included Python 2.7 and 3.2+).
#'      If you need Unicode argument support then you must use Python 3.0+.
#'      Default will be to use \code{findpython} package to find suitable Python binary.
#' @return  \code{ArgumentParser} returns a parser object which contains
#'    an \code{add_argument} function to add arguments to the parser,
#'    a \code{parse_args} function to parse command line arguments into
#'    a list, a \code{print_help} and \code{print_usage} function to print
#'    usage information.  See code examples, package vignette, 
#'    and corresponding python module for more information on how to use it.
#'    
#' @references Python's \code{argparse} library, which this package is based on,
#'  is described here: \url{http://docs.python.org/library/argparse.html}
#' @section Acknowledgement: 
#'     A big thanks to Martin Diehl for a bug report.
#'      
#' @import jsonlite
#' @import proto
#' @import findpython
#' @export
#' @examples
#'
#' parser <- ArgumentParser(description='Process some integers')
#' parser$add_argument('integers', metavar='N', type="integer", nargs='+',
#'                    help='an integer for the accumulator')
#' parser$add_argument('--sum', dest='accumulate', action='store_const',
#'                    const='sum', default='max',
#'                    help='sum the integers (default: find the max)')
#' parser$print_help()
#' # default args for ArgumentParser()$parse_args are commandArgs(TRUE)
#' # which is what you'd want for an Rscript but not for interactive use
#' args <- parser$parse_args(c("--sum", "1", "2", "3")) 
#' accumulate_fn <- get(args$accumulate)
#' print(accumulate_fn(args$integers))
## ifelse(.Platform$OS.type == "windows", "python.exe", "python")
ArgumentParser <- function(..., python_cmd=NULL) {
    python_cmd <- .find_python_cmd(python_cmd)
    .assert_python_cmd(python_cmd) 
    python_code = c("import argparse",
    "try:",
    "    import json",
    "except ImportError:",
    "    import simplejson as json", 
    "",
    sprintf("parser = argparse.ArgumentParser(%s)", 
            convert_..._to_arguments("ArgumentParser", ...)),
    "")
    proto(expr = {
        python_code = python_code
        parse_args = function(., args=commandArgs(TRUE)) {
            python_code <- c(.$python_code, 
                    sprintf("args = parser.parse_args([%s])",
                            paste(sprintf("'%s'", args), collapse=", ")),
                    "print(json.dumps(args.__dict__, sort_keys=True))")
            output <- suppressWarnings(system2(python_cmd,
                        input=python_code, stdout=TRUE, stderr=TRUE))
            if(grepl("^usage:", output[1])) {
                has_positional_arguments <- any(grepl("^positional arguments:", output))
                has_optional_arguments <- any(grepl("^optional arguments:", output))
                if (has_positional_arguments || has_optional_arguments) {
                    .print_message_and_exit(output, "help requested:")
                } else {
                    .stop(output, "parse error:")
                }
            } else if(grepl("^Traceback", output[1])) {
                .stop(output, "Error: python error")
            } else if (grepl("^SyntaxError: Non-ASCII character", output[2])) {
                message <- paste("Non-ASCII character detected.",
                               "If you wish to use Unicode arguments/options",
                               "please upgrade to Python 3.2+",
                               "Please see file INSTALL for more details.")
                .stop(message, "non-ascii character error:")
            } else if (grepl("^SyntaxError: positional argument follows keyword argument", output[2]) || 
                       grepl("^SyntaxError: non-keyword arg after keyword arg", output[2])) {
                message <- paste("Positional argument following keyword argument.",
                                 "Please note ``ArgumentParser`` only accepts keyword arguments.")
                .stop(message, "syntax error:")
            } else if (grepl("^\\{", output)) {
                args <- jsonlite::fromJSON(paste(output, collapse=""))
                return (args)
            } else { # presumably version number request
                .print_message_and_exit(output, "version requested:")
            }
        }
        print_help <- function(.) {
            python_code <- c(.$python_code, "parser.print_help()")
            cat(system(python_cmd, input=python_code, intern=TRUE), sep="\n")   
            invisible(NULL)
        }
        print_usage <- function(.) {
            python_code <- c(.$python_code, "parser.print_usage()")
            cat(system(python_cmd, input=python_code, intern=TRUE), sep="\n")   
            invisible(NULL)
        }
        add_argument = function(., ...) {
            .$python_code <- c(.$python_code,
                    sprintf("parser.add_argument(%s)",
                            convert_..._to_arguments("add_argument", ...)))
            return(invisible(NULL))
        }
    })
}

# @param argument argument to be converted from R to Python
convert_argument <- function(argument) {
    if(is.character(argument)) argument <- shQuote(argument, type="sh") 
    if(is.logical(argument)) argument <- ifelse(argument, 'True', 'False') 
    if(is.null(argument)) argument <- 'None'
    if(length(argument) > 1) {
        argument <- sprintf("(%s)", paste(argument, collapse=", "))
    }
    argument
}

# @param mode Either "add_argument" or "ArgumentParser"
#' @import getopt
convert_..._to_arguments <- function(mode, ...) {

    argument_list <- list(...)
    argument_names <- names(argument_list)
    equals <- ifelse(argument_names == "", "", "=")
    arguments <- shQuote(as.character(argument_list), type="sh")
    proposed_arguments <- paste(argument_names, equals, arguments, sep="")
    # Make sure types are what Python wants
    if(mode == "add_argument" && any(grepl("type=", proposed_arguments))) {
        ii <- grep("type=", proposed_arguments)
        type <- argument_list[[ii]]
        python_type <- switch(type,
                character = "str",
                double = "float",
                integer = "int",
                logical = "bool",
                stop(paste(sprintf("type %s not supported,", type),
                        "supported types:",
                        "'logical', 'integer', 'double' or 'character'")))
        proposed_arguments[ii] <- sprintf("type=%s", python_type)
        # warn if type set to "logical" and action set to "store" 
        if (python_type == "bool" && any(grepl("action='store'", proposed_arguments)))
            warning("You almost certainly want to use action='store_true' or action='store_false' instead")
                                 
    }
    # make sure nargs are what python wants
    if(mode == "add_argument" && any(grepl("nargs=", proposed_arguments))) {
        ii <- grep("nargs=", proposed_arguments)
        nargs <- argument_list[[ii]]
        if(is.numeric(nargs)) {
            nargs <- as.character(nargs)
        } else {
            nargs <- shQuote(nargs, type="sh")
        }
        proposed_arguments[ii] <- sprintf("nargs=%s", nargs)
    }
    if(mode == "add_argument" && any(grepl("choices=", proposed_arguments))) {
        ii <- grep("choices=", proposed_arguments)
        choices <- convert_argument(argument_list[[ii]])
        proposed_arguments[ii] <- sprintf("choices=%s", choices)
    }
    # Fix bug reported by Claire D McWhite
    if(mode == "add_argument" && any(grepl("required=", proposed_arguments))) {
        ii <- grep("required=", proposed_arguments)
        required <- convert_argument(argument_list[[ii]])
        proposed_arguments[ii] <- sprintf("required=%s", required)
    }
    # Feature request from Paul Newell
    if(mode == "add_argument" && any(grepl("metavar=", proposed_arguments))) {
        ii <- grep("metavar=", proposed_arguments)
        metavar <- convert_argument(argument_list[[ii]])
        proposed_arguments[ii] <- sprintf("metavar=%s", metavar)
    }
    # Make defaults are what Python wants, if specified
    default_string <- switch(mode,
           add_argument = "default=", 
           ArgumentParser = "argument_default=",
           stop(sprintf("Unknown mode %s", mode)))
    if(any(grepl(default_string, proposed_arguments))) {
        ii <- grep(default_string, proposed_arguments)
        default <- argument_list[[ii]]
        default <- convert_argument(default)
        proposed_arguments[ii] <- sprintf("%s%s", default_string, default)
    }
    # Don't put quotes around formatter_class argument
    if(mode == "ArgumentParser" && any(grepl("formatter_class=", proposed_arguments))) {
        ii <- grep("formatter_class=", proposed_arguments)
        formatter_class <- argument_list[[ii]]
        proposed_arguments[ii] <- sprintf("formatter_class=%s", formatter_class)
    }
    # Set right default prog name if not specified, if possible
    # Do last to not screw up other fixes with prog insertion
    if(mode == "ArgumentParser" && all(!grepl("prog=", proposed_arguments))) {
        prog <- get_Rscript_filename()
        if(is.na(prog)) prog <- "PROGRAM"
        proposed_arguments <- c(sprintf("prog='%s'", prog), proposed_arguments)
    }
    return(paste(proposed_arguments, collapse=", "))
}

# Internal function to check python cmd is okay
# @param python_cmd Python cmd to use
.assert_python_cmd <- function(python_cmd) {
    if(!is_python_sufficient(python_cmd, required_modules = c('argparse', 'json | simplejson'))) {
        stop(paste(sprintf("python executable %s either is not installed,", python_cmd), 
                "is not on the path, or does not have argparse, json modules",
                "please see INSTALL file"))
    }
}

# Internal function to find python cmd
# @param python_cmd  Python cmd to use
.find_python_cmd <- function(python_cmd) {
    if(is.null(python_cmd)) {
        python_cmd <- getOption("python_cmd")
    }
    if(is.null(python_cmd)) {
        required_modules <- c('argparse', 'json | simplejson')
        did_find_python3 <- can_find_python_cmd(minimum_version='3.0', 
                                                required_modules=required_modules,
                                                silent=TRUE)
        if(did_find_python3) {
            python_cmd <- attr(did_find_python3, "python_cmd")
        } else {
            python_cmd <- find_python_cmd(required_modules=required_modules)
        }
    }
    python_cmd
}

.stop <- function(message, r_note) {
        stop(paste(r_note, paste(message, collapse="\n"), sep="\n"))
}

# Internal function to print message
.print_message_and_exit <- function(message, r_note, status=0) {
    if (interactive()) {
        .stop(message, r_note)
    } else {
        cat(message, sep="\n")
        quit(status=0)
    }
}
